source/common/router/router.cc

```diff
@@ -520,6 +520,21 @@ void Filter::onUpstreamHeaders(Http::HeaderMapPtr&& headers, bool end_stream) {
       upstream_request_->upstream_host_->canary();
   chargeUpstreamCode(*headers, upstream_request_->upstream_host_);
 
+  // Append routing cookies
+  for (auto& cookie : downstream_set_cookies_) {
+    const std::string& key = std::get<0>(cookie);
+    const std::string& value = std::get<1>(cookie);
+    const int max_age = std::get<2>(cookie);
+
+    // Don't overwrite a cookie if it's already set by the upstream host
+    if (Http::Utility::hasSetCookie(*headers, key)) {
```

Comment:
I think the best that can be done right now is to add a perf TODO here. This can be dealt with by potentially working on the TODO in header_map_impl which would allow set-cookie to become an O(1) header but still support append/multiples. I have thoughts on how to do this but probabably out of scope for this PR.